fix: hard fail on invalid railpack.json

iloveitaly opened the pull request:
I noticed this while debugging my project's build failure. I kept getting a message about malformed `railpack.json` even though I did not provide one directly. It seems as through the generated `railpack.json` was incorrect but it did not cause the build to hard fail, it continued on without the generated `railpack.json`.

This change ensures that if invalid railpack.json is encountered, the CLI hard fails.

## Breaking Change

Previously, if JSON did not conform to the schema it would just be ignored and the build would continue. With this change, the build will halt and force the user to fix or discard their schema. I prefer this more explicit approach.

iloveitaly committed:
fix: hard fail if railpack.json cannot be parsed

iloveitaly committed:
test: add shouldFail test.json expression

helpful if we expect the build to fail. There shouldn't be too many of these, but we'll need at least
a couple to make sure cases like railpack.json syntax errors hard fail.

iloveitaly committed:
test: bad json syntax failure

coffee-cup reviewed (commented):
What "generated railpack.json" file are you talking about? I kind of think that we should still be attempting to build if the users config file is malformed. Though could go either way

iloveitaly committed:
test: config file abs path, relative, and DNE cases

iloveitaly committed:
refactor: move hujson helper into internal helpers

iloveitaly committed:
fix: support abs config file path, fail if config file path does not exist

iloveitaly commented:
> What "generated railpack.json" file are you talking about?

Ok, I discovered what was happening here:

In the build script I was testing, I had a `railpack prepare` call which dumped railpack.json. It seems like that command is resulting in invalid `railpack.json`. I was attempting to run `railpack build` directly, which I assumed would ignore `railpack.json` since I didn't specify it as a `--config-file` but it looks like it's read automatically.

Automatically consuming `railpack.json` without a warning feels wrong to me. I think we should only read this during `build` if specified explicitly. Wdyt?

> I kind of think that we should still be attempting to build if the users config file is malformed

I disagree here:

1. If a `railpack.json` exists it was either written by the user or `railpack prepare`. If it's invalid, it's probably a user error.
2. Right now there's only a warning that the schema is invalid and unless you are watching it really quickly it quickly gets gobbled up by a bunch of plan info printing and docker build progress.
3. If a dev breaks `railpack.json` accidentally and a build continues, it could cause a container to get deployed with a build configuration that was not intended, which could result in the wrong start script running, site to break, etc.

---

Two other issues I found here:

1. If an abs path is specified in `--config-file` it always fails since the file path is assumed to relative to the application source. This bit me since the build script I was using was passing an abs path.
2. If the passed config file DNE it is ignored, similar to a railpack.json that can't be parsed

I've updated the PR to correct these. My general thinking here is the provider/detection logic should be "magical" but any user-specified configuration inputs should be strict to guarantee deterministic build plans.

coffee-cup reviewed (commented):
> Automatically consuming railpack.json without a warning feels wrong to me. I think we should only read this during build if specified explicitly. Wdyt?

Can you expand on why it feels wrong? It is very standard across all tooling that supports a config file (Dockerfile, package.json, vite config, buildpacks).

Failing on an invalid `railpack.json` file makes sense then. Agree it will help avoid confusion. [In your case](https://github.com/railwayapp/railpack/issues/204#issuecomment-3086111605) I recommend naming the file `railpack-plan.json`

iloveitaly commented:
> Can you expand on why it feels wrong? It is very standard across all tooling that supports a config file (Dockerfile, package.json, vite config, buildpacks).

On second thought, I'm wrong here. Most build tooling _does_ automatically consume expected files, but most of the other CLI tools I use on a daily basis don't, which is why this felt wrong to me.

I think it's worth adding `path to config file to use (default: railpack.json)` to help output to make this more explicit.

coffee-cup commented:
> I think it's worth adding path to config file to use (default: railpack.json) to help output to make this more explicit.

That would make it much clearer!

iloveitaly committed:
fix: document relative config file path, no abs config file support

iloveitaly committed:
fix: invalid config json

iloveitaly committed:
test: snapshot update

iloveitaly committed:
test: remove abs file config test

iloveitaly committed:
test: remove custom java config

I know basically nothing about Java and after I fixed the syntax error which was causing this config file to not be picked up, this started failing. The proper Zulu version check which is the main purpose of this is still working fine so I think it's fine to remove these custom commands.

iloveitaly committed:
fix: typo

iloveitaly committed:
test: snapshot update

coffee-cup commented on core/core.go:
```diff
@@ -140,17 +164,27 @@ func GenerateConfigFromFile(app *app.App, env *app.Environment, options *Generat
+	// unless an absolute path, assume relative to the app source directory
+	var absConfigFileName string
+	if !filepath.IsAbs(configFileName) {
+		absConfigFileName = filepath.Join(app.Source, configFileName)
```
The issue is that a platform (like Railway) may not want the user to be able to access files outside the app directory. We can't assume that the code being built also has access to all files on the host.

coffee-cup commented on core/core.go:
```diff
@@ -140,17 +164,23 @@ func GenerateConfigFromFile(app *app.App, env *app.Environment, options *Generat
 
-	if !app.HasMatch(configFileName) {
+	// always assume config file path is relative to the app source directory
+	// https://github.com/railwayapp/railpack/pull/226
```
I don't think we need to include a link to the issue/pr for every change like this

theolujay replied:
I see the point made with not leaving link to issue/pr for every change like this. I'd also like to say that I did find this particular one informative as I'm learning the codebase.
